README.md
![Universal Geocoder Logo](https://raw.githubusercontent.com/universal-geocoder/universal-geocoder-js/main/assets/logo.svg)

Universal Geocoder
==================

[![CI](https://img.shields.io/github/workflow/status/universal-geocoder/universal-geocoder-js/Continuous%20Integration?event=push)](https://github.com/universal-geocoder/universal-geocoder-js/actions)
[![codecov](https://img.shields.io/codecov/c/gh/universal-geocoder/universal-geocoder-js/main)](https://codecov.io/gh/universal-geocoder/universal-geocoder-js)

[![npm](https://img.shields.io/npm/v/universal-geocoder)](https://www.npmjs.com/package/universal-geocoder)
[![downloads](https://img.shields.io/npm/dw/universal-geocoder)](https://www.npmjs.com/package/universal-geocoder)

[![dependency-count](https://badgen.net/bundlephobia/dependency-count/universal-geocoder)](https://bundlephobia.com/result?p=universal-geocoder)
[![minified-size](https://img.shields.io/bundlephobia/min/universal-geocoder)](https://bundlephobia.com/result?p=universal-geocoder)
[![tree-shaking](https://badgen.net/bundlephobia/tree-shaking/universal-geocoder)](https://bundlephobia.com/result?p=universal-geocoder)
![types](https://badgen.net/npm/types/universal-geocoder)

Universal Geocoder is a universal JavaScript library for server-side and client-side geocoding applications with multiple built-in providers.

Need geocoding 🌍️ in your website or application? Don't want to be vendor-locked to a service? ✨️ Universal Geocoder ✨️ is for you!

Depending of the chosen provider, it can use geocoding, reverse geocoding or IP geolocation.

Universal Geocoder is a TypeScript fork of [GeocoderJS](https://github.com/geocoder-php/geocoder-js), itself a port of the [Geocoder PHP](https://geocoder-php.org/) library.

This library is platform agnostic: it is available either server-side (**Node**) or client-side (**browsers**, **React Native**, **Electron**).

It aims to be compatible with a maximum of browsers (even old ones) and provides multiple ways to use it, from an old use (available in global environment, results from callbacks) to the most modern use (available as a module, async / await results).

Installation
------------

Add the library to your project:

```shell
npm install --save universal-geocoder
```

> ⚠️ **Warning**: If you need to use this library in an environment not supporting the Promise API such as Internet Explorer, you must install an ES6 Promise compatible polyfill like [es6-promise](https://github.com/jakearchibald/es6-promise).

Usage
-----

You can either use Universal Geocoder as a module (both CommonJS and ES module syntaxes are provided) or as a direct dependency.

As a module:

```javascript
import UniversalGeocoder from "universal-geocoder";

const openStreetMapGeocoder = UniversalGeocoder.createGeocoder("openstreetmap");

// async / await syntax
(async () =>
  console.log(await openStreetMapGeocoder.geocode("1600 Pennsylvania Ave NW, Washington, DC"))
)();

// callback syntax
openStreetMapGeocoder.geocode("1600 Pennsylvania Ave NW, Washington, DC", (result) => {
  console.log(result);
});
```

For this example, the output will be something like this:

```javascript
[
  NominatimGeocoded {
    coordinates: { latitude: 38.8976998, longitude: -77.03655348862276 },
    bounds: {
      latitudeSW: 38.8974898,
      longitudeSW: -77.0368542,
      latitudeNE: 38.897911,
      longitudeNE: -77.0362526
    },
    formattedAddress: undefined,
    streetNumber: '1600',
    streetName: 'Pennsylvania Avenue Northwest',
    subLocality: undefined,
    locality: 'Washington',
    postalCode: '20500',
    region: 'District of Columbia',
    adminLevels: [
      AdminLevel {
        level: 1,
        name: 'District of Columbia',
        code: undefined
      },
      AdminLevel {
        level: 2,
        name: 'Washington',
        code: undefined
      }
    ],
    country: 'United States of America',
    countryCode: 'us',
    timezone: undefined,
    displayName: 'White House, 1600, Pennsylvania Avenue Northwest, Washington, District of Columbia, 20500, United States of America',
    osmId: 238241022,
    osmType: 'way',
    category: 'historic',
    type: 'castle',
    attribution: 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright'
  },
  // ... (other results)
]
```

If you want to use the library as a direct dependecy (for browsers only), copy `dist/universal-geocoder.js` or `dist/universal-geocoder.min.js` to your dependencies.

Universal Geocoder will be available in the global environment:

```javascript
const openStreetMapGeocoder = UniversalGeocoder.createGeocoder("openstreetmap");

openStreetMapGeocoder.geodecode("44.915", "-93.21", (result) => {
  console.log(result);
});
```

For a more advanced usage, see the example below:

```javascript
import UniversalGeocoder, { ReverseQuery } from "universal-geocoder";

const googleGeocoder = UniversalGeocoder.createGeocoder({
  provider: "googlemaps",
  apiKey: "YOUR_API_KEY",
  useSsl: true,
  useJsonp: false,
  // other specific provider options
});

(async () =>
  console.log(await googleGeocoder.geocode({
    text: "1600 Pennsylvania Ave, Washington, DC",
    locale: "FR",
    limit: 10,
    // other specific provider parameters
  }))
)();

const reverseQuery = ReverseQuery.create({
  coordinates: {
    latitude: "44.915",
    longitude: "-93.21",
  },
})
.withLocale("FR")
.withLimit(7);
(async () =>
  console.log(await googleGeocoder.geodecode(reverseQuery))
)();
```

### Common Options (`createGeocoder` method)

- `useSsl`: boolean to use the HTTPS API of the providers
- `useJsonp`: boolean to use JSONP
- `apiKey`: the API key to use for the provider

### Common `geocode` parameters (`GeocodeQuery` object)

- `text`: what is searched
- `ip`: the IP searched
- `bounds` (object with `latitudeSW`, `longitudeSW`, `latitudeNE` and `longitudeNE` keys): the bounds to use (either bias or filter the results)
- `locale`: the locale to use for the query
- `limit`: the maximum number of results to have

### Common `geodecode` parameters (`ReverseQuery` object)

- `coordinates` (object with `latitude`, `longitude` keys): the coordinates to search for
- `locale`: the locale to use for the query
- `limit`: the maximum number of results to have

### Common Result Properties (`Geocoded` object)

The result of a query is a `Geocoded` object which maps the following common information:
- Coordinates (object with `latitute` and `longitude` keys)
- Bounds (object with `latitudeSW`, `longitudeSW`, `latitudeNE`, `longitudeNE` keys)
- Formatted address
- Address details: street number, street name, (sub) locality, postal code, region, administration levels, country (with its code)
- Time zone

You can either use getter methods to retrieve them or use the `toObject` method to manipulate an object containing the properties.

Providers
---------

Universal Geocoder comes with modules to integrate with various geocoding providers.

The following table summarizes the features of each:

<table>
  <thead>
    <tr>
      <th>Provider</th>
      <th>Codename</th>
      <th>Supports location geocoding?</th>
      <th>Supports reverse geocoding?</th>
      <th>Supports IP geolocation?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>OpenStreetMap (Nominatim)</td>
      <td>openstreetmap or nominatim</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>OpenCage</td>
      <td>opencage</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>Google Maps (Geocoding API)</td>
      <td>google or googlemaps</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>LocationIQ</td>
      <td>locationiq</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>Mapbox</td>
      <td>mapbox</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>MapQuest</td>
      <td>mapquest</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>Bing</td>
      <td>bing</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>Yandex</td>
      <td>yandex</td>
      <td>✅️ yes</td>
      <td>✅️ yes</td>
      <td>❌️ no</td>
    </tr>
    <tr>
      <td>GeoPlugin</td>
      <td>geoplugin</td>
      <td>❌️ no</td>
      <td>❌️ no</td>
      <td>✅️ yes</td>
    </tr>
  </tbody>
</table>

### Specific Provider Usage

The documentation for specific provider options, parameters and results can be found [here](docs/provider_usage.md).

Special Providers
-----------------

A `chain` provider is available: it iterates over multiple providers.

For more information, see [its documentation](docs/provider_usage/chain.md).

Dumpers
-------

Dumpers transform a `Geocoded` object to another format.

### GeoJSON

[GeoJSON](https://geojson.org/) is a format for encoding a variety of geographic data structures.

#### Usage

```javascript
import UniversalGeocoder, { GeoJsonDumper } from "universal-geocoder";

const nominatimGeocoder = UniversalGeocoder.createGeocoder("nominatim");

(async () => {
  const result = await nominatimGeocoder.geocode("1600 Pennsylvania Ave NW, Washington, DC");
  console.log(result);
  console.log("GeoJSON:", GeoJsonDumper.dump(result[0]));
})();
```

Building
--------

```shell
npm run build
```

Testing
-------

Unit and functional tests are handled by Jasmine. To run tests from the command line, use:

```shell
npm test
```

If you need to record new API calls, use:

```shell
npm run test-record
```

You can also check if the examples are running correctly.

For the Web:

```shell
npm run serve
```

Then go to http://localhost:8080/example/web, choose a provider and open the console.

For Node:

```shell
npm run ts-node -- example/node/provider.ts
```

FAQ
---

**Q: When using Universal Geocoder client-side, how to make sure the API key is not stolen?**

**A:** First of all, there are some providers that do not use an API key, like OpenStreetMap (Nominatim) or GeoPlugin.

If you want to use a provider with an API key, the best approach is generally to use Universal Geocoder in the server-side (Node) and to call it from the client-side.
This way the API key is not exposed directly.

Some providers allow to add URL restrictions when generating a token, like [Mapbox](https://docs.mapbox.com/help/getting-started/access-tokens/#url-restrictions).
Even if the token is visible, it mitigates its unwanted use since the `Origin` header [cannot be changed](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name) in a browser environment.
However, the stolen token would be still usable in a server environment (for instance with cURL).


CHANGELOG.md
# Changelog

## 0.14.1

* Export missing types

## 0.14.0

* Better TypeScript inference for `createGeocoder`, `geocode` and `geodecode`

## 0.13.0

* [Nominatim] Add `shape` and `shapeThreshold` geocode and geodecode parameters
* [Nominatim] Add `shape` in geocoded object

## 0.12.0

* Add LocationIQ provider
* [Nominatim] Add `countryCodes` to provider options
* [Nominatim] Add `dedupe` geocode parameter
* [Nominatim] Replace `type` with `types` (now an array) in geocoded object
* [Nominatim] Replace `category` with `categories` (now an array) in geocoded object
* [Mapbox] Replace `geocodingMode` with `source` in provider options
* [Mapbox] Replace `resultType` with `types` in geocoded object
* [Mapbox] Replace `locationTypes` with `types` for geocode and geodecode
* [MapQuest] Replace `openDomain` with `source` in provider options
* [GoogleMaps] Replace `locationType` with `precision` in geocoded object
* [GoogleMaps] Replace `resultType` with `types` in geocoded object
* [GoogleMaps] Replace `locationTypes` with `precisions` for geodecode
* [GoogleMaps] Replace `resultTypes` with `types` for geodecode
* [Yandex] Replace `locationType` with `types` (now an array) in geocoded object
* [Yandex] Replace `locationTypes` with `types` for geodecode

## 0.11.1

* Module is side-effect free

## 0.11.0

* Export ES module as well as CommonJS

## 0.10.0

* [Nominatim] Add `subLocalityLevels`

## 0.9.0

* Introduce async / await (or Promise) syntax (providers return a promise if no callback is given)

## 0.8.0

* [Bing] Better Bing provider
* [Mapbox] Add `fuzzyMatch` geocode parameter
* [OpenCage] Add `precision` to `OpenCageGeocoded`
* [OpenCage] Replace `minConfidence` with `minPrecision` for geocode and geodecode

## 0.7.0

* [Yandex] Better Yandex provider
* [Nominatim] Remove `viewBox` parameter in `GeocodeQuery` to use `bounds` instead
* [GeoJsonDumper] Use `bbox` member for `bounds`
* The `bounds` parameter in `GeocodeQuery` is now an object with `latitudeSW`, `longitudeSW`, `latitudeNE`, `longitudeNE` keys
* The `latitude` and `longitude` parameters in `GeocodeQuery` are replaced by a `coordinates` parameter (an object with `latitude` and `longitude` keys)
* `Geocoded` returns an object with `latitudeSW`, `longitudeSW`, `latitudeNE`, `longitudeNE` keys for the `bounds`
* The `latitude` and `longitude` parameters in `Geocoded` are replaced by a `coordinates` parameter (an object with `latitude` and `longitude` keys)

## 0.6.0

* [MapQuest] Better MapQuest provider

## 0.5.0

* Add special chain provider
* Add GeoPlugin provider
* Add `errorCallback` parameter to `geocode` and `geodecode`
* Add `body` parameter to `executeRequest`
* [MapQuest] Fix geocode request (bad encoded query)

## 0.4.0

* [Google] Better Google Maps provider
* [Nominatim] Better handling of geodecode errors
* [ExternalLoader] Add `getOptions` method to interface
* [ExternalLoader] Rename `JSONPCallback` in params to `jsonpCallback`
* Rename `GoogleAPIProvider` to `GoogleMapsProvider`
* Add Node examples

## 0.3.0

* Add OpenCage Provider
* Add time zone to `Geocoded`
* [Mapbox] Add `countryCodes`, `proximity` and `reverseMode` to geo(de)code parameters
* [Mapbox] Add `resultType` to `Geocoded`
* [Mapbox] Add admin levels to `Geocoded`
* Rename `ExternalURILoader` to `ExternalLoader`

## 0.2.4

* [Nominatim] Fix error when no result is received

## 0.2.3

* Fix TypeScript types for the library to be usable in a TS project

## 0.2.2

* Use relative paths for build files

## 0.2.1

* Missing types folder when publishing

## 0.2.0

* Better Nominatim support
* Better GeoJSON support

## 0.1.0

* Library rewritten in TypeScript
* Library can be used as a module as well
* Add Mapbox provider
* Formatted address, bounds and country support
* Better handling of provider options (JSONP, SSL)
* The methods `geocode` and `geodecode` can take an object or a `Query`


example/node/mapquest.ts
/* eslint-disable no-console */
import UniversalGeocoder from "UniversalGeocoder";
import Geocoded from "Geocoded";
import { MapQuestGeocodeQuery, MapQuestLocation } from "provider";

let mapQuestGeocoder = UniversalGeocoder.createGeocoder({
  provider: "mapquest",
  apiKey: "api_key",
});

mapQuestGeocoder.geocode(
  "1600 Pennsylvania Ave NW, Washington, DC",
  (result) => {
    console.log("geocode", result);
  }
);
const asyncGeocode = async () => {
  const result = await mapQuestGeocoder.geocode(
    "1600 Pennsylvania Ave NW, Washington, DC"
  );
  console.log("async geocode", result);
};
asyncGeocode();

mapQuestGeocoder.geodecode("44.915", "-93.21", (result) => {
  console.log("geodecode", result);
});
const asyncGeodecode = async () => {
  const result = await mapQuestGeocoder.geodecode("44.915", "-93.21");
  console.log("async geodecode", result);
};
asyncGeodecode();

mapQuestGeocoder.geocode(
  MapQuestGeocodeQuery.create({
    location: Geocoded.create({
      streetNumber: "1600",
      streetName: "Pennsylvania Ave NW",
      locality: "Washington",
    }),
  }),
  (result) => {
    console.log("geocode location", result);
  }
);

mapQuestGeocoder = UniversalGeocoder.createGeocoder({
  provider: "mapquest",
  apiKey: "Fmjtd|luurnu6al1,bg=o5-9wbg94",
  method: "POST",
  source: "nominatim",
});

mapQuestGeocoder.geocode(
  "1600 Pennsylvania Ave NW, Washington, DC",
  (result) => {
    console.log("geocode POST", result);
  }
);

mapQuestGeocoder.geodecode("44.915", "-93.21", (result) => {
  console.log("geodecode POST", result);
});

mapQuestGeocoder.geocode(
  MapQuestGeocodeQuery.create({
    location: MapQuestLocation.create({
      street: "1600 Pennsylvania Ave NW",
      city: "Washington",
    }),
  }),
  (result) => {
    console.log("geocode POST location", result);
  }
);
